queue_services/entity-bn/src/entity_bn/bn_processors/change_of_registration.py

```diff
@@ -0,0 +1,191 @@
+from entity_bn.exceptions import BNException
+
+
+def process(business: Business, filing: Filing):  # pylint: disable=too-many-branches
+    """Process the incoming change of registration request."""
+    if not business.tax_id or len(business.tax_id) != 15:
+        raise BNException(f'Business {business.identifier}, ' +
+                          'Cannot inform CRA about change of registration before receiving Business Number (BN15).')
+
+    if filing.meta_data and filing.meta_data.get('changeOfRegistration', {}).get('toLegalName'):
+        _change_name(business, filing, RequestTracker.RequestType.CHANGE_NAME)
+
+    with suppress(KeyError, ValueError):
+        if dpath.util.get(filing.filing_json, 'filing/changeOfRegistration/offices/businessOffice'):
+            if has_previous_address(filing.transaction_id,
+                                    business.delivery_address.one_or_none().office_id, 'delivery'):
+                _change_address(business, filing, RequestTracker.RequestType.CHANGE_DELIVERY_ADDRESS)
+
+            if has_previous_address(filing.transaction_id,
+                                    business.mailing_address.one_or_none().office_id, 'mailing'):
+                _change_address(business, filing, RequestTracker.RequestType.CHANGE_MAILING_ADDRESS)
+
+
+def _change_name(business: Business, filing: Filing,  # pylint: disable=too-many-locals
+                 name_type: RequestTracker.RequestType):
+    """Inform CRA about change of name."""
+    max_retry = current_app.config.get('BN_HUB_MAX_RETRY')
+    request_trackers = RequestTracker.find_by(business.id,
+                                              RequestTracker.ServiceName.BN_HUB,
+                                              RequestTracker.RequestType.CHANGE_NAME,
+                                              filing.id)
+    if not request_trackers:
+        request_tracker = RequestTracker()
+        request_tracker.business_id = business.id
+        request_tracker.filing_id = filing.id
+        request_tracker.request_type = RequestTracker.RequestType.CHANGE_NAME
+        request_tracker.service_name = RequestTracker.ServiceName.BN_HUB
+    elif (request_tracker := request_trackers.pop()) and not request_tracker.is_processed:
+        request_tracker.last_modified = datetime.utcnow()
+        request_tracker.retry_number += 1
+
+    if request_tracker.is_processed:
+        return
+
+    business_registration_number = business.tax_id[0:9]
+    business_program_identifier = business.tax_id[9:11]
+    business_program_account_reference_number = business.tax_id[11:15]
+
+    client_name_type_code = {
+        # RequestTracker.RequestType.CHANGE_PARTY: '01',
+        RequestTracker.RequestType.CHANGE_NAME: '02'
+    }
+    update_reason_code = {
+        # RequestTracker.RequestType.CHANGE_PARTY: '03',
+        RequestTracker.RequestType.CHANGE_NAME: '01'
+    }
+
+    new_name = business.legal_name if name_type == RequestTracker.RequestType.CHANGE_NAME else ''
```

Comment:
Will the incoming `name_type` always be `RequestTracker.RequestType.CHANGE_NAME`?

Reply:
When Partner/Proprietor changes we have to send the same xml input with some differences in values. You might have noticed RequestType.CHANGE_PARTY (yet to be implemented), will be the other name_type.